text-editor/text-editor-002-en.md

```diff
@@ -0,0 +1,227 @@
+The drawback of this approach is that we're working directly on the view, so
+we do not have a model representation anymore that we could render. The view
+was our model.
+
+Once more, the MDN was our friend for this study :
+- the [DOM Node API](https://developer.mozilla.org/en-US/docs/Web/API/Node)
+permitted us to navigate through the node tree
+- we were able to [get the selected text](https://developer.mozilla.org/en-US/docs/Web/API/Selection)
+(returns a [range](https://developer.mozilla.org/en-US/docs/Web/API/Range))
+- we were simply creating elements to insert with
+[`document.createElement`](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement)
+
+We had to be very careful with the DOM manipulation as some Node methods are
+directly mutating the node itself. We should update the node only we have
+finished to compute the changes to apply. During this computation, we had
+to make copies of some node so we could edit them without chabge the actual
+view.
+
+We managed to have a working system for text mutations (bold, italic) across
+a single or multiple paragraphs, even when the selection was overlapsing
+multiple HTML tags (see the above example).
+
+The hard part was to produce the cleanest HTML markup as possible, eg for a
+`Hello world` text, when `Hello ` is already bold and we make `world` bold
+too, we initialy had
+```html
+<b>Hello </b><b>world</b>
+```
+two `b` tags following each other, but this should have been simplified to
+```html
+<b>Hello world</b>
+```
+This sanitization was done in a pass once the new `b` tag was added to the DOM.
+
+In the end, this was working fine but the number of lines of code started to
+grow, and we were discovering more and more edge cases which were painful to
+cover.
+
+Finally, we discovered a method which would have saved us all.
+
+### The holy grail : document.execCommand
+
+And then we discovered
+[`document.execCommand`](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand) \o/
+This function applies some text transformations to the current selection. Eg
+to bold / unbold the selection, just execute :
+```js
+document.execCommand('bold')
+```
+and that's it ! Your selection is wrapped inside a `b` tag, the produced
+markup is very clean (remember the previous step where we had to sanitize the
+markup manually).
+This is very neat, it enables rich text edition directly in the browser (and
+by the browser :p).
+Checkout [all the available commands](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand#Commands)
+to see what's possible to do !
+Needless to say that when we discovered this, we immediately stopped the
+previous attempt which was about manipulating the DOM by ourselves, and
+it was definitely time for a beer of victory !
+
+Considering the commands list, we had all the operators we needed to build
+our text editor. We only had to execute the right command when clicking on
+the right button.
+
```

Comment:
I think we can write a little paragraph here to explain that some cases remained hard to handle (ie preserve html structure when inserting a paragraph after an inserted figure such as a tweet or an image). WDYT ?